pkg/skaffold/build/cache.go

```diff
@@ -0,0 +1,373 @@
+	if cacheFile != "" {
+		return cacheFile, util.VerifyOrCreateFile(cacheFile)
+	}
+	home, err := homedir.Dir()
+	if err != nil {
+		return "", errors.Wrap(err, "retrieving home directory")
+	}
+	defaultFile := filepath.Join(home, constants.DefaultSkaffoldDir, constants.DefaultCacheFile)
+	return defaultFile, util.VerifyOrCreateFile(defaultFile)
+}
+
+func retrieveArtifactCache(cacheFile string) (ArtifactCache, error) {
+	cache := ArtifactCache{}
+	contents, err := ioutil.ReadFile(cacheFile)
+	if err != nil {
+		return nil, err
+	}
+	if err := yaml.Unmarshal(contents, &cache); err != nil {
+		return nil, err
+	}
+	return cache, nil
+}
+
+// RetrieveCachedArtifacts checks to see if artifacts are cached, and returns tags for cached images, otherwise a list of images to be built
+func (c *Cache) RetrieveCachedArtifacts(ctx context.Context, out io.Writer, artifacts []*latest.Artifact) ([]*latest.Artifact, []Artifact) {
+	if !c.useCache {
+		return artifacts, nil
+	}
+	color.Default.Fprintln(out, "Checking cache...")
+	var needToBuild []*latest.Artifact
+	var built []Artifact
+	for _, a := range artifacts {
+		artifact, err := c.retrieveCachedArtifact(ctx, out, a)
+		if err != nil {
+			logrus.Debugf("error retrieving cached artifact for %s: %v\n", a.ImageName, err)
+			needToBuild = append(needToBuild, a)
+			continue
+		}
+		if artifact == nil {
+			needToBuild = append(needToBuild, a)
+			continue
+		}
+		built = append(built, *artifact)
+	}
+	return needToBuild, built
+}
+
+func (c *Cache) retrieveCachedArtifact(ctx context.Context, out io.Writer, a *latest.Artifact) (*Artifact, error) {
+	hash, err := hashForArtifact(ctx, a)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting hash for artifact %s", a.ImageName)
+	}
+	a.WorkspaceHash = hash
+	imageDetails, cacheHit := c.artifactCache[hash]
+	if !cacheHit {
+		return nil, nil
+	}
+	newTag := fmt.Sprintf("%s:%s", a.ImageName, hash)
+	// Check if tagged image exists remotely with the same digest
+	existsRemotely := imageExistsRemotely(newTag, imageDetails.Digest)
+	// Check if we are using a local cluster
+	local, _ := localCluster()
+
+	// See if this image exists in the local daemon
+	if c.client.ImageExists(ctx, newTag) {
+		color.Yellow.Fprintf(out, "Found %s locally...\n", a.ImageName)
+		// Push if the image doesn't exist remotely
+		if !existsRemotely && !local {
+			color.Yellow.Fprintf(out, "Pushing %s since it doesn't exist remotely...\n", a.ImageName)
+			if _, err := c.client.Push(ctx, out, newTag); err != nil {
+				return nil, errors.Wrapf(err, "pushing %s", newTag)
+			}
+		}
+		color.Yellow.Fprintf(out, "%s ready, skipping rebuild\n", newTag)
+		return &Artifact{
+			ImageName: a.ImageName,
+			Tag:       newTag,
+		}, nil
+	}
+	// Check for a local image with the same digest as the image we want to build
+	prebuiltImage, err := c.retrievePrebuiltImage(ctx, imageDetails)
+	if err != nil {
+		return nil, errors.Wrapf(err, "getting prebuilt image")
+	}
+	if prebuiltImage == "" {
```

Comment:
can this ever happen?

Reply:
I think so, while testing there were cases when I could match to an image by digest but there was no associated tagged image (which is what we need for retagging).